dd-java-agent/instrumentation/servlet/src/main/java/datadog/trace/instrumentation/servlet/request/HttpServletRequestCallSite.java

```
@@ -0,0 +1,94 @@
+      }
+    }
+    return headerValue;
+  }
+
+  @CallSite.AfterArray({
+    @CallSite.After(
+        "java.util.Enumeration javax.servlet.http.HttpServletRequest.getHeaders(java.lang.String)"),
+    @CallSite.After(
+        "java.util.Enumeration javax.servlet.http.HttpServletRequestWrapper.getHeaders(java.lang.String)"),
+  })
+  public static Enumeration<?> afterGetHeaders(
+      @CallSite.This final HttpServletRequest self,
+      @CallSite.Argument final String headerName,
+      @CallSite.Return final Enumeration<?> enumeration) {
+    if (enumeration == null) {
+      return null;
+    }
+    final WebModule module = InstrumentationBridge.WEB;
+    if (module == null) {
+      return enumeration;
+    }
+    final List<String> headerValues = new ArrayList<>();
+    while (enumeration.hasMoreElements()) {
+      final String headerValue = (String) enumeration.nextElement();
+      headerValues.add(headerValue);
+      try {
+        module.onHeaderValue(headerName, headerValue);
+      } catch (final Throwable e) {
+        module.onUnexpectedException("afterGetHeaders threw", e);
+      }
+    }
+    return Collections.enumeration(headerValues);
+  }
+
+  @CallSite.AfterArray({
+    @CallSite.After("java.util.Enumeration javax.servlet.http.HttpServletRequest.getHeaderNames()"),
+    @CallSite.After(
+        "java.util.Enumeration javax.servlet.http.HttpServletRequestWrapper.getHeaderNames()"),
+  })
+  public static Enumeration<?> afterGetHeaderNames(
+      @CallSite.This final HttpServletRequest self,
+      @CallSite.Return final Enumeration<?> enumeration) {
+    if (enumeration == null) {
+      return null;
+    }
+    final WebModule module = InstrumentationBridge.WEB;
+    if (module == null) {
+      return enumeration;
+    }
+    final List<String> headerNames = new ArrayList<>();
+    while (enumeration.hasMoreElements()) {
+      final String headerName = (String) enumeration.nextElement();
```

This should probably be within the try/catch block?

---

Maybe also the `hasMoreElements()`...

---

Nice catch, it's a bit more complex like move it withing the try/catch block. If they fail it's servlet code that fails and we should let the exception propagate removing ourselves from the stack first. Check the latest version and the specific tests for that matter:

```java
  def 'test that get headers does not fail when servlet related code fails'(final Class<? extends HttpServletRequest> clazz) {
    setup:
    final iastModule = Mock(WebModule)
    InstrumentationBridge.registerIastModule(iastModule)
    final enumeration = Mock(Enumeration) {
      hasMoreElements() >> { throw new NuclearBomb('Boom!!!')}
    }
    final testSuite = new TestHttpServletRequestCallSiteSuite(Mock(clazz) {
      getHeaders('header') >> enumeration
    })

    when:
    testSuite.getHeaders('header')

    then:
    final bomb = thrown(NuclearBomb)
    bomb.stackTrace.find { it.className == HttpServletRequestCallSite.name } == null

    where:
    clazz                     | _
    HttpServletRequest        | _
    HttpServletRequestWrapper | _
  }
```